Remove material drawer depenency

timrae opened the pull request:
In v2.4 we "manually" implemented the navigation drawer, only using the `DrawerLayout` view provided in the support library. This was quite messy and hard to extend, so when the design support library was released I switched over to the new `NavigationView` class, which simplified the code and allowed us to add the header image to the top of the drawer, to give a more "material" theme to the app, and increase Anki branding.

In the end we wanted to add a switch for night mode, and `NavigationView` didn't support this. So eventually we switched over to the 3rd party library "material drawer", which allowed custom views. This simplified life somewhat, but it also led to a few hidden bugs that were really painful to solve.

Google ended up releasing a new version of `NavigationView` which included support for custom views (like switches) shortly after we switched to the 3rd party library (lol), and this PR switches back to an implementation based on that class again, and removes our dependency on the material drawer library.

Fullscreen mode is currently broken; I'm waiting on some feedback from Google on what I think is a bug before I try to hack a solution (obviously I'm not holding my breath for their reply, but I'll fix it eventually one way or another). Another issue is that the colors in night mode are a bit different  from 2.5 also, I'll try to take a look at that later as well when I get time.

timrae commented:
if someone could at least test the keyboard related stuff that would be useful

ospalh commented:
This crashed for my as soon as i start it.
(I did bump the version strings from beta2 to beta22 to avoid an uninstall-reinstall.)

```
E/AndroidRuntime(24085): FATAL EXCEPTION: main
E/AndroidRuntime(24085): Process: com.ichi2.anki, PID: 24085
E/AndroidRuntime(24085): java.lang.IllegalArgumentException: DrawerLayout must be measured with MeasureSpec.EXACTLY.
E/AndroidRuntime(24085): 	at android.support.v4.widget.DrawerLayout.onMeasure(DrawerLayout.java:903)
E/AndroidRuntime(24085): 	at android.view.View.measure(View.java:17562)
E/AndroidRuntime(24085): 	at android.widget.LinearLayout.measureHorizontal(LinearLayout.java:1063)
E/AndroidRuntime(24085): 	at android.widget.LinearLayout.onMeasure(LinearLayout.java:615)
E/AndroidRuntime(24085): 	at android.view.View.measure(View.java:17562)
E/AndroidRuntime(24085): 	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:5536)
E/AndroidRuntime(24085): 	at android.widget.FrameLayout.onMeasure(FrameLayout.java:436)
E/AndroidRuntime(24085): 	at android.support.v7.widget.ContentFrameLayout.onMeasure(ContentFrameLayout.java:135)
E/AndroidRuntime(24085): 	at android.view.View.measure(View.java:17562)
E/AndroidRuntime(24085): 	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:5536)
E/AndroidRuntime(24085): 	at android.widget.FrameLayout.onMeasure(FrameLayout.java:436)
E/AndroidRuntime(24085): 	at android.view.View.measure(View.java:17562)
E/AndroidRuntime(24085): 	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:5536)
E/AndroidRuntime(24085): 	at android.widget.FrameLayout.onMeasure(FrameLayout.java:436)
E/AndroidRuntime(24085): 	at android.view.View.measure(View.java:17562)
E/AndroidRuntime(24085): 	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:5536)
E/AndroidRuntime(24085): 	at android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1436)
E/AndroidRuntime(24085): 	at android.widget.LinearLayout.measureVertical(LinearLayout.java:722)
E/AndroidRuntime(24085): 	at android.widget.LinearLayout.onMeasure(LinearLayout.java:613)
E/AndroidRuntime(24085): 	at android.view.View.measure(View.java:17562)
E/AndroidRuntime(24085): 	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:5536)
E/AndroidRuntime(24085): 	at android.widget.FrameLayout.onMeasure(FrameLayout.java:436)
E/AndroidRuntime(24085): 	at com.android.internal.policy.impl.PhoneWindow$DecorView.onMeasure(PhoneWindow.java:2871)
E/AndroidRuntime(24085): 	at android.view.View.measure(View.java:17562)
E/AndroidRuntime(24085): 	at android.view.ViewRootImpl.performMeasure(ViewRootImpl.java:2015)
E/AndroidRuntime(24085): 	at android.view.ViewRootImpl.measureHierarchy(ViewRootImpl.java:1173)
E/AndroidRuntime(24085): 	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:1379)
E/AndroidRuntime(24085): 	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1061)
E/AndroidRuntime(24085): 	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:5891)
E/AndroidRuntime(24085): 	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:767)
E/AndroidRuntime(24085): 	at android.view.Choreographer.doCallbacks(Choreographer.java:580)
E/AndroidRuntime(24085): 	at android.view.Choreographer.doFrame(Choreographer.java:550)
E/AndroidRuntime(24085): 	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:753)
E/AndroidRuntime(24085): 	at android.os.Handler.handleCallback(Handler.java:739)
E/AndroidRuntime(24085): 	at android.os.Handler.dispatchMessage(Handler.java:95)
E/AndroidRuntime(24085): 	at android.os.Looper.loop(Looper.java:135)
E/AndroidRuntime(24085): 	at android.app.ActivityThread.main(ActivityThread.java:5292)
E/AndroidRuntime(24085): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(24085): 	at java.lang.reflect.Method.invoke(Method.java:372)
E/AndroidRuntime(24085): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:904)
E/AndroidRuntime(24085): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:699)
W/ActivityManager(  676):   Force finishing activity 1 com.ichi2.anki/.DeckPicker
```

With 
AnkiDroid Version = 2.5beta22

Android Version = 5.1.1 (CGM)

ACRA UUID = beb26812-2a73-4b04-96d1-fd67daf954ce

hssm commented:
I also got that crash and couldn't test it.

You still need to add back the `android:windowSoftInputMode="adjustResize"` that we talked about in the manifest to make the soft keyabord resize thing work.

timrae commented:
OK I've fixed that crash on tablets and added `android:windowSoftInputMode="adjustResize"` to the manifest. Fullscreen is broken; I'm waiting on some feedback from Google before I try to hack a solution. The colors are a bit different also, but this change generates a lot of conflicts so I think a quick merge is useful.

hssm commented:
Did we have the fullscreen working correctly at one point with the sticky immersive mode? I know you probably won't agree with me but I liked that mode more than the current one because it didn't resize the content which I'm finding a little annoying. I gave up on that mode because of the problems with showing the action bar in overlay mode but with the removal of that nav drawer library we might be able to get it to work now.

>Let's not put anymore non-critical changes into the 2.5.x branch.

Yeah I agree. I imagine 2.6 won't stay in development as long as 2.5 did.

timrae commented:
> Did we have the fullscreen working correctly at one point with the sticky immersive mode?

The problem with sticky immersive is that you don't get a callback when the status bar is shown, so you can't control show/hide of other elements like the action bar (which the docs say you have to hide together with the status bar). 

> I liked (sticky immersive) mode more than the current one because it didn't resize the content which I'm finding a little annoying.

That's an implementation issue rather than something inherent in the sticky vs non-sticky discussion. Actually in the sample app (see [this video](http://www.youtube.com/watch?v=cBi8fjv90E4&t=2m30s)) they don't resize the content, and we could make AnkiDroid work that way as well, which I think is a good idea.

> with the removal of that nav drawer library we might be able to get it to work now

I eventually found that our fullscreen issues come down to a kind of [bug](https://code.google.com/p/android/issues/detail?id=195813) in `DrawerLayout` from the support library. In the nav drawer library (which also uses that API), I think he's put in a few workarounds that that add confusion because they don't work perfectly, but at the end of the day we probably need to wait for an upstream fix.

hssm commented:
Okay I understand now. I realized where my confusion was coming from. When we used the sticky immersive mode, the system/status bar had a transparent theme applied when they were displayed, so in my mind I thought that a transparent action bar (like in the YouTube app) was also the result of sticky immersive.

So avoiding resizing content is what overlay mode does, and the transparency is part of the theme, and those are both down to the implementation. That all makes sense now.
